server/sonar-server/src/main/java/org/sonar/server/qualityprofile/BuiltInQProfileUpdateImpl.java

```diff
@@ -45,22 +47,31 @@ public BuiltInQProfileUpdateImpl(DbClient dbClient, RuleActivator ruleActivator,
     this.activeRuleIndexer = activeRuleIndexer;
   }
 
-  public List<ActiveRuleChange> update(DbSession dbSession, BuiltInQProfile builtIn, RulesProfileDto ruleProfile) {
+  public List<ActiveRuleChange> update(DbSession dbSession, BuiltInQProfile builtIn, RulesProfileDto rulesProfile) {
     // Keep reference to all the activated rules before update
-    Set<RuleKey> toBeDeactivated = dbClient.activeRuleDao().selectByRuleProfile(dbSession, ruleProfile)
+    Set<RuleKey> deactivatedKeys = dbClient.activeRuleDao().selectByRuleProfile(dbSession, rulesProfile)
       .stream()
       .map(ActiveRuleDto::getRuleKey)
       .collect(MoreCollectors.toHashSet());
 
-    List<ActiveRuleChange> changes = new ArrayList<>();
-    builtIn.getActiveRules().forEach(ar -> {
+    Collection<RuleActivation> activations = new ArrayList<>();
+    Collection<RuleKey> ruleKeys = new HashSet<>(deactivatedKeys);
+    for (BuiltInActiveRule ar : builtIn.getActiveRules()) {
       RuleActivation activation = convert(ar);
-      toBeDeactivated.remove(activation.getRuleKey());
-      changes.addAll(ruleActivator.activateOnBuiltInRulesProfile(dbSession, activation, ruleProfile));
-    });
+      activations.add(activation);
+      ruleKeys.add(activation.getRuleKey());
+      deactivatedKeys.remove(activation.getRuleKey());
+    }
+
+    RuleActivationContext context = ruleActivator.createContextForBuiltInProfile(dbSession, rulesProfile, ruleKeys);
+
+    List<ActiveRuleChange> changes = new ArrayList<>();
```

Comment:
old school! you can write this with a flatmap :D

Reply:
apparently, you could use flatmap in several locations in this commit

Reply:
I will check, thanks